packages/account-tree-controller/src/backup-and-sync/service/index.ts

```diff
@@ -0,0 +1,554 @@
+    // Set up cleanup without affecting the returned promise
+    promise
+      .finally(() => {
+        this.#ongoingFullSyncPromise = null;
+      })
+      .catch(() => {
+        // Only ignore errors from the cleanup operation itself
+        // The original promise errors are still propagated to callers
+      });
+    return promise;
+  }
+
+  /**
+   * Enqueues a single wallet sync operation (fire-and-forget).
+   *
+   * @param walletId - The wallet ID to sync.
+   */
+  enqueueSingleWalletSync(walletId: AccountWalletId): void {
+    if (!this.isBackupAndSyncEnabled) {
+      return;
+    }
+
+    if (!this.hasSyncedAtLeastOnce) {
+      // Run big sync
+      // eslint-disable-next-line no-void
+      void this.performFullSync();
+      return;
+    }
+    // eslint-disable-next-line no-void
+    void this.#atomicSyncQueue.enqueue(() =>
+      this.#performSingleWalletSyncInner(walletId),
+    );
+  }
+
+  /**
+   * Enqueues a single group sync operation (fire-and-forget).
+   *
+   * @param groupId - The group ID to sync.
+   */
+  enqueueSingleGroupSync(groupId: AccountGroupId): void {
+    if (!this.isBackupAndSyncEnabled) {
+      return;
+    }
+
+    if (!this.hasSyncedAtLeastOnce) {
+      // Run big sync
+      // eslint-disable-next-line no-void
+      void this.performFullSync();
+      return;
+    }
+
+    // eslint-disable-next-line no-void
+    void this.#atomicSyncQueue.enqueue(() =>
+      this.#performSingleGroupSyncInner(groupId),
+    );
+  }
+
+  /**
+   * Performs a full synchronization of the local account tree with user storage, ensuring consistency
+   * between local state and cloud-stored account data.
+   * If a full sync is already in progress, it will return the ongoing promise.
+   * This clears the atomic sync queue before starting the full sync.
+   *
+   * NOTE: in some very edge cases, this can be ran concurrently if triggered quickly after
+   * toggling back and forth the backup and sync feature from the UI.
+   *
+   * @returns A promise that resolves when the sync is complete.
+   */
+  async performFullSync(): Promise<void> {
+    if (!this.isBackupAndSyncEnabled) {
+      return undefined;
+    }
+
+    // If there's an ongoing sync (including first sync), return it
+    if (this.#ongoingFullSyncPromise) {
+      return this.#ongoingFullSyncPromise;
+    }
+
+    // First sync setup - create and cache the first sync promise
+    if (!this.#firstOngoingFullSyncPromise) {
+      this.#firstOngoingFullSyncPromise = this.#atomicSyncQueue.clearAndEnqueue(
+        () => this.#performFullSyncInner(),
+      );
+      return this.#setupOngoingPromiseCleanup(
+        this.#firstOngoingFullSyncPromise,
+      );
+    }
+
+    // Create a new ongoing sync (sequential calls after previous completed)
+    const newSyncPromise = this.#atomicSyncQueue.clearAndEnqueue(() =>
+      this.#performFullSyncInner(),
+    );
+
+    return this.#setupOngoingPromiseCleanup(newSyncPromise);
```

Comment:
Nit: But I think we could have use this instead:

```suggestion
    // Create a new ongoing sync (sequential calls after previous completed)
    const newSyncPromise = this.#atomicSyncQueue.clearAndEnqueue(() =>
      this.#performFullSyncInner(),
    );

    // First sync setup - create and cache the first sync promise
    if (!this.#firstOngoingFullSyncPromise) {
      this.#firstOngoingFullSyncPromise = newSyncPromise;
    }

    return this.#setupOngoingPromiseCleanup(newSyncPromise);
```

This should be same, it's just that `#firstOngoingFullSyncPromise` will hold a reference to the very first `newSyncPromise` until its cleared.